File: leetcode/235-lowest-common-ancestor-of-a-binary-search-tree/lowest_common_ancestor.py
""" 
LC 235: Lowest Common Ancestor of a Binary Search Tree
"""
class Solution:
    def lowest_common_ancestor(self, root: Optional[TreeNode], p: Optional[TreeNode], q: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root or not p or not q:
            return None
        if (max(p.val, q.val) < root.val):
            return self.lowest_common_ancestor(root.left, p, q)
        elif (min(p.val, q.val) > root.val):
            return self.lowest_common_ancestor(root.right, p, q)
        else:
            return root

File: leetcode/235-lowest-common-ancestor-of-a-binary-search-tree/LowestCommonAncestor.java
/* 
 *  LC 235: Lowest Common Ancestor of a Binary Search Tree
* */

public class LowestCommonAncestor {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(int val) {
            this.val = val;
        }
    }

    private TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || p == null || q == null)
            return null;
        
        if (Math.max(p.val, q.val) < root.val)
            return lowestCommonAncestor(root.left, p, q);
        else if (Math.min(p.val, q.val) > root.val)
            return lowestCommonAncestor(root.right, p, q);
        else
            return root;
    }
}